src/cunumeric/nullary/window_util.h

```diff
@@ -0,0 +1,126 @@
+ *
+ */
+
+#pragma once
+
+#define _USE_MATH_DEFINES
+
+#include "cunumeric/cunumeric.h"
+#include <math.h>
+
+extern double i0(double);
+
+namespace cunumeric {
+
+enum class WindowOpCode : int {
+  BARLETT  = CUNUMERIC_WINDOW_BARLETT,
+  BLACKMAN = CUNUMERIC_WINDOW_BLACKMAN,
+  HAMMING  = CUNUMERIC_WINDOW_HAMMING,
+  HANNING  = CUNUMERIC_WINDOW_HANNING,
+  KAISER   = CUNUMERIC_WINDOW_KAISER,
+};
+
+template <typename Functor, typename... Fnargs>
+constexpr decltype(auto) op_dispatch(WindowOpCode op_code, Functor f, Fnargs&&... args)
+{
+  switch (op_code) {
+    case WindowOpCode::BARLETT:
+      return f.template operator()<WindowOpCode::BARLETT>(std::forward<Fnargs>(args)...);
+    case WindowOpCode::BLACKMAN:
+      return f.template operator()<WindowOpCode::BLACKMAN>(std::forward<Fnargs>(args)...);
+    case WindowOpCode::HAMMING:
+      return f.template operator()<WindowOpCode::HAMMING>(std::forward<Fnargs>(args)...);
+    case WindowOpCode::HANNING:
+      return f.template operator()<WindowOpCode::HANNING>(std::forward<Fnargs>(args)...);
+    case WindowOpCode::KAISER:
+      return f.template operator()<WindowOpCode::KAISER>(std::forward<Fnargs>(args)...);
+  }
+  assert(false);
+  return f.template operator()<WindowOpCode::BARLETT>(std::forward<Fnargs>(args)...);
+}
+
+template <WindowOpCode OP_CODE>
+struct WindowOp;
+
+template <>
+struct WindowOp<WindowOpCode::BARLETT> {
+  WindowOp(int64_t M, double beta) : alpha_(static_cast<double>(M - 1) / 2.0) {}
+  constexpr double operator()(int64_t idx) const
+  {
+    return idx < alpha_ ? idx / alpha_ : 2.0 - idx / alpha_;
+  }
+
+  double alpha_;
+};
+
+template <>
+struct WindowOp<WindowOpCode::BLACKMAN> {
+  WindowOp(int64_t M, double beta) : alpha_(M_PI * 2 / (M - 1)) {}
+  constexpr double operator()(int64_t idx) const
+  {
+    using std::cos;
+    double val = idx * alpha_;
+    return 0.42 - 0.5 * cos(val) + 0.08 * cos(2 * val);
+  }
+
+  double alpha_;
+};
+
+template <>
+struct WindowOp<WindowOpCode::HAMMING> {
+  WindowOp(int64_t M, double beta) : alpha_(M_PI * 2 / (M - 1)) {}
+  constexpr double operator()(int64_t idx) const { return 0.54 - 0.46 * std::cos(idx * alpha_); }
+
+  double alpha_;
+};
+
+template <>
+struct WindowOp<WindowOpCode::HANNING> {
+  WindowOp(int64_t M, double beta) : alpha_(M_PI * 2 / (M - 1)) {}
+  constexpr double operator()(int64_t idx) const { return 0.5 - 0.5 * std::cos(idx * alpha_); }
+
+  double alpha_;
+};
+
+template <>
+struct WindowOp<WindowOpCode::KAISER> {
+  WindowOp(int64_t M, double beta) : alpha_((M - 1) / 2.0), beta_(beta) {}
+
+#if defined(__NVCC__) || defined(__CUDACC__)
+  __device__ double operator()(int64_t idx) const
+  {
+    using std::sqrt;
+    auto val    = (idx - alpha_) / alpha_;
+    auto result = cyl_bessel_i0(beta_ * sqrt(1 - val * val));
+    result /= cyl_bessel_i0(beta_);
+    return result;
+  }
+#else
+  double operator()(int64_t idx) const
+  {
+    using std::sqrt;
+    auto val    = (idx - alpha_) / alpha_;
```

Comment:
should  val be `val = idx/alpha_`

Reply:
The code I have here is what NumPy and CuPy does. `idx / alpha_` is greater than 1 for the second half of the array and makes the subsequent `sqrt` call return a nan.

Reply:
I see, thanks